py4j-python/src/py4j/clientserver.py

```diff
@@ -203,6 +203,32 @@ class JavaClient(GatewayClient):
     ClientServerConnection per thread.
     """
 
+    class ThreadLocalConnectionCleaner(object):
+        """Cleans :class:`ClientServerConnection` in the thread local with
+        closing it properly, and removes it from the deque at
+        :class:`JavaClient`. Right before the Python thread is terminated,
+        this instance is garbage-collected with triggering __del__ that
+        contains a cleaning-up logic to prevent a resource leak.
+        """
+        def __init__(self, connection, dequeue):
+            assert (
+                isinstance(connection, Callable) and
+                connection() is not None and
+                isinstance(connection(), ClientServerConnection))
+            self.connection = connection
+            self.deque = dequeue
+
+        def __del__(self):
+            conn = self.connection()
```

Comment:
I would add a docstring.

```
"""Removes the connection associated with the current thread from the deque.

Expected to be called when the thread that started the connection is garbage-collected.
"""
```

The second sentence was what I struggled to understand while reviewing the code.